### ISSUE TITLE
Not very clear documentation

### ISSUE THREAD

**Issue by ezhikus** [microsoft/wil#97]
In your documentation you say: "To get started testing WIL, first make sure that you have a recent version of Visual Studio and the most recent Windows SDK installed."

I use Visual Studio 2017 and Win10 SDK - and library can't be compiled because of some unknown attributes in header files. Looks like VS 2017 and/or Win10 SDK is not enough?

Could you please mention in documentation exact version of VS and SDK which are required to use WIL ?

**Comment by sylveon** [comment]
What are the errors?

**Comment by ezhikus** [comment]
It doesn't know what is "_Translates_last_error_to_HRESULT_". As far as I understand this thing is defined in some versions of Win10 SDK and is not defined in others. That's why would be great to write in doc which version of Win10 SDK is required.

**Comment by sylveon** [comment]
That macro is defined in `specstrings.h` but WIL does not include this header directly. Does it still happens if you manually include this header before including WIL?

**Comment by ezhikus** [comment]
Yes, but only in some versions of Win10 SDK. For example, Win 10 SDK 10.0.10240.0 doesn't include this file. I've copy-pasted this macro definition from Win10 SDK 10.0.17134.0 - and now it works fine for me. 
But it is strange you use some macro without explicit include of external header file where it is defined. And I still think it would be good to define exact versions of Visual Studio and Win10 SDK which should be used for WIL.

**Comment by sylveon** [comment]
I personally assume it requires a recent-ish version since that library is used for internal Windows development.

**Comment by dunhor** [comment]
A few thoughts/comments. The first is that the "recent version" comment was primarily targeted towards building/running the tests. Really it should probably say "most recent version" but that's not entirely true - older versions _might_ work and we've been broken by newer versions as well. The tests, in contrast to the library itself, expect a certain degree of functionality being present (e.g. as opposed to excluding tests that can't be compiled using the SDK being targeted on the machine).

The story for WIL - the library - is a bit different. Historically it began and evolved in the OS repo (and continues to some degree). This means that, at least for a while, the assumption was that the "target SDK" was the latest and greatest and a pretty recent version of the compiler - typically very close to the most recent public version of the compiler. That is to say that new features in WIL could introduce dependencies on new features in the OS with minimal repercussions. That slowly somewhat changed as WIL started to get copied to other internal projects (and eventually to its own internal NuGet package) that build outside of the OS. Thus this earlier assumption broke and compatibility with older compilers and SDKs started to improve.

But there's never been - to my knowledge - any initiative to maintain any "minimum" compatibility. In general, updates and new additions should be cognizant of any new dependencies they add and conditionally exclude these additions as appropriate. But sometimes this doesn't work out perfectly - sometimes people make mistakes, sometimes a feature has been around long enough that nobody notices it was added more recently than everyone thought, and sometimes the "minimum" is bumped - even if not necessarily intentionally - to simplify code (e.g. variable template support is assumed throughout, effectively setting the minimum C++ version to C++14). Most of the time these things are fixable, but it's generally been the responsibility of those who are using WIL in these manners to identify and fix - or find someone to fix - these problems as they arise.

Not a perfect solution - there's certainly room to do better here. The obvious thing is to modify the build pipeline to test more scenarios, but that's not necessarily as good or easy of an idea as it might seem. The build already takes quite a while to run and the availability/control of compiler/SDK/etc. with Azure DevOps is not as great as one would like it to be, at least not the last time I played with it. Another option - one that's much more feasible that I might look into - is for me to run a build on an older SDK before publishing the NuGet package (I've already started to wait for a successful Windows build before publishing). No promises that issues will get fixed then and there, but at the very least we can have bugs filed for things that don't work.

So if you would like a minimum "this is guaranteed to work" version, I'd suggest using what the CI build uses (or if that's not easy to look up, the most recent public releases), but by no means should you take that to mean that's the absolute minimum. Feel free to file bugs - or even better, fixes - for issues that you encounter. E.g. in this particular scenario it seems like it would be rather simple to conditionally provide a definition for `_Translates_last_error_to_HRESULT_` when not otherwise available (i.e. similar to what you've done locally)
